ddtrace/contrib/kafka_python_ng/patch.py

```diff
@@ -0,0 +1,231 @@
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.schema import schematize_messaging_operation
+from ddtrace.internal.schema import schematize_service_name
+from ddtrace.internal.schema.span_attribute_schema import SpanDirection
+from ddtrace.internal.utils import ArgumentError
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.utils import set_argument_value
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.pin import Pin
+from ddtrace.propagation.http import HTTPPropagator as Propagator
+
+
+_KafkaProducer = kafka.KafkaProducer
+_KafkaConsumer = kafka.KafkaConsumer
+
+log = get_logger(__name__)
+
+config._add(
+    "kafka",
+    dict(
+        _default_service=schematize_service_name("kafka"),
+        distributed_tracing_enabled=asbool(os.getenv("DD_KAFKA_PROPAGATION_ENABLED", default=False)),
+        trace_empty_poll_enabled=asbool(os.getenv("DD_KAFKA_EMPTY_POLL_ENABLED", default=True)),
+    ),
+)
+
+
+def get_version():
+    return getattr(kafka, "__version__", "")
+
+
+class TracedKafkaProducerMixin:
+    def __init__(self, **config):
+        super(TracedKafkaProducerMixin, self).__init__(**config)
+        self._dd_bootstrap_servers = (
+            config.get("bootstrap_servers")
+            if config.get("bootstrap_servers") is not None
+            else config.get("metadata_broker_list")
+        )
+
+
+class TracedKafkaConsumerMixin:
+    def __init__(self, *topics, **configs):
+        super(TracedKafkaConsumerMixin, self).__init__(*topics, **configs)
+        self._group_id = configs.get("group_id", "")
+        self._auto_commit = asbool(configs.get("enable_auto_commit", True))
+
+
+class TracedKafkaConsumer(TracedKafkaConsumerMixin, kafka.KafkaConsumer):
+    pass
+
+
+class TracedKafkaProducer(TracedKafkaProducerMixin, kafka.KafkaProducer):
+    pass
+
+
+def patch():
+    if getattr(kafka, "_datadog_patch", False):
+        return
+
+    kafka.KafkaProducer = TracedKafkaProducer
+    kafka.KafkaConsumer = TracedKafkaConsumer
+
+    trace_utils.wrap(TracedKafkaProducer, "send", traced_send)
+    trace_utils.wrap(TracedKafkaConsumer, "poll", traced_poll)
+    trace_utils.wrap(TracedKafkaConsumer, "commit", traced_commit)
+
+    Pin().onto(kafka.KafkaProducer)
+    Pin().onto(kafka.KafkaConsumer)
+
+    kafka._datadog_patch = True
+
+
+def unpatch():
+    if trace_utils.iswrapped(TracedKafkaProducer.send):
+        trace_utils.unwrap(TracedKafkaProducer, "send")
+    if trace_utils.iswrapped(TracedKafkaConsumer.poll):
+        trace_utils.unwrap(TracedKafkaConsumer, "poll")
+    if trace_utils.iswrapped(TracedKafkaConsumer.commit):
+        trace_utils.unwrap(TracedKafkaConsumer, "commit")
+
+    kafka.KafkaProducer = _KafkaProducer
+    kafka.KafkaConsumer = _KafkaConsumer
+
+    if getattr(kafka, "_datadog_patch", False):
+        kafka._datadog_patch = False
+
+
+def traced_send(func, instance, args, kwargs):
```

Comment:
Rather than copypasting from `contrib.kafka`, I recommend putting some of that modules code into utility functions in a new file `utils_kafka.py` and importing them from both `contrib.kafka` and `contrib.kafka_python_ng`.

Reply:
I'll see what is possible, it's not exact copy paste, there's differences in method signatures and data structures.
It could turn out to be a mini abstraction class for both kafka libs.